<?php
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: google/genomics/v1/annotations.proto

namespace Google\Genomics\V1;

use Google\Protobuf\Internal\GPBType;
use Google\Protobuf\Internal\RepeatedField;
use Google\Protobuf\Internal\GPBUtil;

/**
 * <pre>
 * An annotation describes a region of reference genome. The value of an
 * annotation may be one of several canonical types, supplemented by arbitrary
 * info tags. An annotation is not inherently associated with a specific
 * sample or individual (though a client could choose to use annotations in
 * this way). Example canonical annotation types are `GENE` and
 * `VARIANT`.
 * </pre>
 *
 * Protobuf type <code>google.genomics.v1.Annotation</code>
 */
class Annotation extends \Google\Protobuf\Internal\Message
{
    /**
     * <pre>
     * The server-generated annotation ID, unique across all annotations.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    private $id = '';
    /**
     * <pre>
     * The annotation set to which this annotation belongs.
     * </pre>
     *
     * <code>string annotation_set_id = 2;</code>
     */
    private $annotation_set_id = '';
    /**
     * <pre>
     * The display name of this annotation.
     * </pre>
     *
     * <code>string name = 3;</code>
     */
    private $name = '';
    /**
     * <pre>
     * The ID of the Google Genomics reference associated with this range.
     * </pre>
     *
     * <code>string reference_id = 4;</code>
     */
    private $reference_id = '';
    /**
     * <pre>
     * The display name corresponding to the reference specified by
     * `referenceId`, for example `chr1`, `1`, or `chrX`.
     * </pre>
     *
     * <code>string reference_name = 5;</code>
     */
    private $reference_name = '';
    /**
     * <pre>
     * The start position of the range on the reference, 0-based inclusive.
     * </pre>
     *
     * <code>int64 start = 6;</code>
     */
    private $start = 0;
    /**
     * <pre>
     * The end position of the range on the reference, 0-based exclusive.
     * </pre>
     *
     * <code>int64 end = 7;</code>
     */
    private $end = 0;
    /**
     * <pre>
     * Whether this range refers to the reverse strand, as opposed to the forward
     * strand. Note that regardless of this field, the start/end position of the
     * range always refer to the forward strand.
     * </pre>
     *
     * <code>bool reverse_strand = 8;</code>
     */
    private $reverse_strand = false;
    /**
     * <pre>
     * The data type for this annotation. Must match the containing annotation
     * set's type.
     * </pre>
     *
     * <code>.google.genomics.v1.AnnotationType type = 9;</code>
     */
    private $type = 0;
    /**
     * <pre>
     * A map of additional read alignment information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    private $info;
    protected $value;

    public function __construct() {
        \GPBMetadata\Google\Genomics\V1\Annotations::initOnce();
        parent::__construct();
    }

    /**
     * <pre>
     * The server-generated annotation ID, unique across all annotations.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * <pre>
     * The server-generated annotation ID, unique across all annotations.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public function setId($var)
    {
        GPBUtil::checkString($var, True);
        $this->id = $var;
    }

    /**
     * <pre>
     * The annotation set to which this annotation belongs.
     * </pre>
     *
     * <code>string annotation_set_id = 2;</code>
     */
    public function getAnnotationSetId()
    {
        return $this->annotation_set_id;
    }

    /**
     * <pre>
     * The annotation set to which this annotation belongs.
     * </pre>
     *
     * <code>string annotation_set_id = 2;</code>
     */
    public function setAnnotationSetId($var)
    {
        GPBUtil::checkString($var, True);
        $this->annotation_set_id = $var;
    }

    /**
     * <pre>
     * The display name of this annotation.
     * </pre>
     *
     * <code>string name = 3;</code>
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * <pre>
     * The display name of this annotation.
     * </pre>
     *
     * <code>string name = 3;</code>
     */
    public function setName($var)
    {
        GPBUtil::checkString($var, True);
        $this->name = $var;
    }

    /**
     * <pre>
     * The ID of the Google Genomics reference associated with this range.
     * </pre>
     *
     * <code>string reference_id = 4;</code>
     */
    public function getReferenceId()
    {
        return $this->reference_id;
    }

    /**
     * <pre>
     * The ID of the Google Genomics reference associated with this range.
     * </pre>
     *
     * <code>string reference_id = 4;</code>
     */
    public function setReferenceId($var)
    {
        GPBUtil::checkString($var, True);
        $this->reference_id = $var;
    }

    /**
     * <pre>
     * The display name corresponding to the reference specified by
     * `referenceId`, for example `chr1`, `1`, or `chrX`.
     * </pre>
     *
     * <code>string reference_name = 5;</code>
     */
    public function getReferenceName()
    {
        return $this->reference_name;
    }

    /**
     * <pre>
     * The display name corresponding to the reference specified by
     * `referenceId`, for example `chr1`, `1`, or `chrX`.
     * </pre>
     *
     * <code>string reference_name = 5;</code>
     */
    public function setReferenceName($var)
    {
        GPBUtil::checkString($var, True);
        $this->reference_name = $var;
    }

    /**
     * <pre>
     * The start position of the range on the reference, 0-based inclusive.
     * </pre>
     *
     * <code>int64 start = 6;</code>
     */
    public function getStart()
    {
        return $this->start;
    }

    /**
     * <pre>
     * The start position of the range on the reference, 0-based inclusive.
     * </pre>
     *
     * <code>int64 start = 6;</code>
     */
    public function setStart($var)
    {
        GPBUtil::checkInt64($var);
        $this->start = $var;
    }

    /**
     * <pre>
     * The end position of the range on the reference, 0-based exclusive.
     * </pre>
     *
     * <code>int64 end = 7;</code>
     */
    public function getEnd()
    {
        return $this->end;
    }

    /**
     * <pre>
     * The end position of the range on the reference, 0-based exclusive.
     * </pre>
     *
     * <code>int64 end = 7;</code>
     */
    public function setEnd($var)
    {
        GPBUtil::checkInt64($var);
        $this->end = $var;
    }

    /**
     * <pre>
     * Whether this range refers to the reverse strand, as opposed to the forward
     * strand. Note that regardless of this field, the start/end position of the
     * range always refer to the forward strand.
     * </pre>
     *
     * <code>bool reverse_strand = 8;</code>
     */
    public function getReverseStrand()
    {
        return $this->reverse_strand;
    }

    /**
     * <pre>
     * Whether this range refers to the reverse strand, as opposed to the forward
     * strand. Note that regardless of this field, the start/end position of the
     * range always refer to the forward strand.
     * </pre>
     *
     * <code>bool reverse_strand = 8;</code>
     */
    public function setReverseStrand($var)
    {
        GPBUtil::checkBool($var);
        $this->reverse_strand = $var;
    }

    /**
     * <pre>
     * The data type for this annotation. Must match the containing annotation
     * set's type.
     * </pre>
     *
     * <code>.google.genomics.v1.AnnotationType type = 9;</code>
     */
    public function getType()
    {
        return $this->type;
    }

    /**
     * <pre>
     * The data type for this annotation. Must match the containing annotation
     * set's type.
     * </pre>
     *
     * <code>.google.genomics.v1.AnnotationType type = 9;</code>
     */
    public function setType($var)
    {
        GPBUtil::checkEnum($var, \Google\Genomics\V1\AnnotationType::class);
        $this->type = $var;
    }

    /**
     * <pre>
     * A variant annotation, which describes the effect of a variant on the
     * genome, the coding sequence, and/or higher level consequences at the
     * organism level e.g. pathogenicity. This field is only set for annotations
     * of type `VARIANT`.
     * </pre>
     *
     * <code>.google.genomics.v1.VariantAnnotation variant = 10;</code>
     */
    public function getVariant()
    {
        return $this->readOneof(10);
    }

    /**
     * <pre>
     * A variant annotation, which describes the effect of a variant on the
     * genome, the coding sequence, and/or higher level consequences at the
     * organism level e.g. pathogenicity. This field is only set for annotations
     * of type `VARIANT`.
     * </pre>
     *
     * <code>.google.genomics.v1.VariantAnnotation variant = 10;</code>
     */
    public function setVariant(&$var)
    {
        GPBUtil::checkMessage($var, \Google\Genomics\V1\VariantAnnotation::class);
        $this->writeOneof(10, $var);
    }

    /**
     * <pre>
     * A transcript value represents the assertion that a particular region of
     * the reference genome may be transcribed as RNA. An alternative splicing
     * pattern would be represented as a separate transcript object. This field
     * is only set for annotations of type `TRANSCRIPT`.
     * </pre>
     *
     * <code>.google.genomics.v1.Transcript transcript = 11;</code>
     */
    public function getTranscript()
    {
        return $this->readOneof(11);
    }

    /**
     * <pre>
     * A transcript value represents the assertion that a particular region of
     * the reference genome may be transcribed as RNA. An alternative splicing
     * pattern would be represented as a separate transcript object. This field
     * is only set for annotations of type `TRANSCRIPT`.
     * </pre>
     *
     * <code>.google.genomics.v1.Transcript transcript = 11;</code>
     */
    public function setTranscript(&$var)
    {
        GPBUtil::checkMessage($var, \Google\Genomics\V1\Transcript::class);
        $this->writeOneof(11, $var);
    }

    /**
     * <pre>
     * A map of additional read alignment information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    public function getInfo()
    {
        return $this->info;
    }

    /**
     * <pre>
     * A map of additional read alignment information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    public function setInfo(&$var)
    {
        $this->info = $var;
    }

    public function getValue()
    {
        return $this->value;
    }

}

